src/goto-programs/remove_virtual_functions.cpp

```diff
@@ -144,6 +146,63 @@ static void create_static_function_call(
   }
 }
 
+/// If \p argument_for_this contains a dereference then create a temporary
+/// variable for it and use that instead
+/// \param [in,out] argument_for_this: The first argument of the function call
+/// \param symbol_table: The symbol table to add the new symbol to
+/// \param vcall_source_loc: The source location of the function call, which is
+///   used for new instructions that are added
+/// \param [out] new_code_for_this_argument: New instructions are added here
+static void process_this_argument(
+  exprt &argument_for_this,
+  symbol_table_baset &symbol_table,
+  const source_locationt &vcall_source_loc,
+  goto_programt &new_code_for_this_argument)
+{
+  const auto dereference_it = std::find_if(
+    argument_for_this.depth_cbegin(),
+    argument_for_this.depth_cend(),
+    [](const exprt &expr) { return can_cast_expr<dereference_exprt>(expr); });
+
+  const bool argument_for_this_contains_a_dereference =
+    dereference_it != argument_for_this.depth_cend();
+
+  if(argument_for_this_contains_a_dereference)
+  {
+    // Create a temporary for the `this` argument. This is so that value-set
+    // filtering in symex will reduce the value-set for `this` to those with the
+    // correct class identifier.
+    symbolt &temp_symbol = get_fresh_aux_symbol(
+      argument_for_this.type(),
+      "remove_virtual_functions", // TODO: pass function name in to use here?
+      "this_argument",
+      vcall_source_loc,
+      ID_java, //TODO: get this from somewhere
```

Comment:
Require that name of the `goto_program` that's being processed is passed in as well (just like `remove_function_pointers` does this, search for `function_id`). You can then look up the symbol corresponding to `function_id` and use the `mode` of that symbol.